backend/src/main/java/io/spring/cloud/samples/animalrescue/backend/AnimalRepository.java

```diff
@@ -1,6 +1,122 @@
 package io.spring.cloud.samples.animalrescue.backend;
 
-import org.springframework.data.repository.CrudRepository;
 
-public interface AnimalRepository extends CrudRepository<Animal, Long> {
-}
+import java.time.LocalDate;
+import java.util.Collection;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.Map;
+import java.util.function.Function;
+
+import io.r2dbc.spi.Row;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+import org.springframework.r2dbc.core.DatabaseClient;
+import org.springframework.r2dbc.core.DatabaseClient.GenericExecuteSpec;
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Transactional;
+import org.springframework.util.CollectionUtils;
+
+@Repository
+public class AnimalRepository {
```

Comment:
I guess the trigger to roll your own repository implementation is because we have nested object `AdoptionRequest` inside `Animal`. Is that right? If so, a potentially more readable and maintainable option is to decouple these two repositories and move the logic to the application code as [suggested here](https://github.com/spring-projects/spring-data-r2dbc/issues/137#issuecomment-500236478). 

For example, getting all animals can be implemented like this:

```
animalRepository.findAll()
                .flatMap(animal -> adoptionRequestRepository.findAllByAnimal(animal.getId())
                                                            .collectList()
                                                            .map(list -> animal.setAdoptionRequests(list)));
```

Reply:
@LittleBaiBai Thanks for the comment.

This code above would cause the "N+1 selects problem" that I tend to avoid.

[The issue](https://github.com/spring-projects/spring-data-r2dbc/issues/137#issuecomment-500236478) you are referencing also says:
> Running multiple SQL statements to retrieve objects is less efficient than running a single statement.

The proposed code runs only one SQL to retrieve all animal objects.
I was thinking `handle resolution of relationships in your application code` means something like [this](https://github.com/spring-cloud-services-samples/animal-rescue/pull/89/files#diff-9f34ee61fa5eb7eea60cf18eb293bf3208f8cd027bf822c105c5b994eaa86866R51-R69).


But I understand your concern. This application is not expected to be used in production, so it makes sense to prioritize readability.
How about adding a warning as follows?

```java
// This code prioritizes being more readable and maintainable 
// and causes the "N+1 selects problem". Take care of referring to the code.
animalRepository.findAll()
                .flatMap(animal -> adoptionRequestRepository.findAllByAnimal(animal.getId())
                                                            .collectList()
                                                            .map(list -> animal.setAdoptionRequests(list)));
```

Reply:
That warning is perfect. Thanks a lot for being flexible!

Reply:
Updated